lib/Jsrt/JsrtDiag.cpp

```diff
@@ -0,0 +1,688 @@
+        {
+            return JsErrorInThreadServiceCallback;
+        }
+        else if (threadContext->IsInScript())
+        {
+            return JsErrorRuntimeInUse;
+        }
+        else if (runtime->GetJsrtDebugManager() != nullptr && runtime->GetJsrtDebugManager()->IsDebugEventCallbackSet())
+        {
+            return JsErrorDiagAlreadyInDebugMode;
+        }
+
+        ThreadContextScope scope(threadContext);
+
+        if (!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+        // Create the debug object to save callback function and data
+        runtime->EnsureJsrtDebugManager();
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        jsrtDebugManager->SetDebugEventCallback(debugEventCallback, callbackState);
+
+        if (threadContext->GetDebugManager() != nullptr)
+        {
+            threadContext->GetDebugManager()->SetLocalsDisplayFlags(Js::DebugManager::LocalsDisplayFlags::LocalsDisplayFlags_NoGroupMethods);
+        }
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            Assert(!scriptContext->IsScriptContextInDebugMode());
+
+            Js::DebugContext* debugContext = scriptContext->GetDebugContext();
+
+            if (debugContext->GetHostDebugContext() == nullptr)
+            {
+                debugContext->SetHostDebugContext(jsrtDebugManager);
+            }
+
+            if (FAILED(scriptContext->OnDebuggerAttached()))
+            {
+                AssertMsg(false, "Failed to start debugging");
+                return JsErrorFatal; // Inconsistent state, we can't continue from here?
+            }
+
+            Js::ProbeContainer* probeContainer = debugContext->GetProbeContainer();
+            probeContainer->InitializeInlineBreakEngine(jsrtDebugManager);
+            probeContainer->InitializeDebuggerScriptOptionCallback(jsrtDebugManager);
+        }
+
+        return JsNoError;
+    });
+}
+
+CHAKRA_API JsDiagStopDebugging(
+    _In_ JsRuntimeHandle runtimeHandle,
+    _Out_ void** callbackState)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        VALIDATE_INCOMING_RUNTIME_HANDLE(runtimeHandle);
+
+        PARAM_NOT_NULL(callbackState);
+
+        *callbackState = nullptr;
+
+        JsrtRuntime * runtime = JsrtRuntime::FromHandle(runtimeHandle);
+        ThreadContext * threadContext = runtime->GetThreadContext();
+
+        if (threadContext->GetRecycler() && threadContext->GetRecycler()->IsHeapEnumInProgress())
+        {
+            return JsErrorHeapEnumInProgress;
+        }
+        else if (threadContext->IsInThreadServiceCallback())
+        {
+            return JsErrorInThreadServiceCallback;
+        }
+        else if (threadContext->IsInScript())
+        {
+            return JsErrorRuntimeInUse;
+        }
```

Comment:
converted to a macro

---
In reply to: [62731016](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62731016) [](ancestors = 62731016)